marlowe-dashboard-client/src/Capability/Marlowe.purs

```diff
@@ -1,77 +1,235 @@
+  -- note: this returns an empty walletNickname (because these are only saved locally) and an empty
+  -- marloweContractId (because this would take a while to lookup - to get the marloweContractId we
+  -- call `marloweGetContracts`)
+  marloweLookupWalletDetails :: ContractInstanceId -> m (AjaxResponse WalletDetails)
+  -- get the observable state of a wallet's WalletCompanionContract
+  marloweGetRoleContracts :: WalletDetails -> m (DecodedAjaxResponse (Map MarloweParams MarloweData))
+  -- get all contracts related to a wallet
+  marloweGetContracts :: WalletDetails -> m (DecodedAjaxResponse (Tuple WalletDetails (Map ContractInstanceId History)))
 
-instance monadMarloweAppM :: ManageMarloweContract AppM where
-  marloweCreateWalletCompanionContract wallet = activateContract $ ContractActivationArgs { caID: walletCompanionContractExe, caWallet: toBack wallet }
-  marloweGetWalletCompanionContractObservableState contractInstanceId = do
+instance monadMarloweAppM :: ManageMarlowe AppM where
+  marloweCreateWallet = do
+    ajaxWalletInfo <- createWallet
+    case ajaxWalletInfo of
+      Left ajaxError -> pure $ Left ajaxError
+      Right walletInfo -> do
+        let
+          wallet = view _wallet walletInfo
+        ajaxMarloweContractId <- activateContract (contractExe MarloweContract) wallet
+        ajaxCompanionContractId <- activateContract (contractExe WalletCompanionContract) wallet
+        ajaxAssets <- getWalletTotalFunds wallet
+        case ajaxMarloweContractId, ajaxCompanionContractId, ajaxAssets of
+          Left ajaxError, _, _ -> pure $ Left ajaxError
+          _, Left ajaxError, _ -> pure $ Left ajaxError
+          _, _, Left ajaxError -> pure $ Left ajaxError
+          Right marloweContractId, Right companionContractId, Right assets ->
+            pure
+              $ Right
+                  { walletNickname: ""
+                  , marloweContractId: marloweContractId
+                  , companionContractId: companionContractId
+                  , walletInfo: walletInfo
+                  , assets
+                  }
+  marloweFollowContract walletDetails marloweParams = do
+    let
+      wallet = view (_walletInfo <<< _wallet) walletDetails
+    ajaxFollowContractId <- activateContract (contractExe WalletFollowerContract) wallet
+    case ajaxFollowContractId of
+      Left ajaxError -> pure $ Left $ Left ajaxError
+      Right followContractId -> do
+        _ <- invokeEndpoint followContractId "follow" marloweParams
+        ajaxObservableState <- getContractInstanceObservableState followContractId
+        case ajaxObservableState of
+          Left ajaxError -> pure $ Left $ Left ajaxError
+          Right rawJson -> case runExcept $ decodeJSON $ unwrap rawJson of
+            Left decodingError -> pure $ Left $ Right decodingError
+            Right observableState -> pure $ Right $ followContractId /\ observableState
```

Comment:
Good use of `runExceptT`, thanks for that. Good idea with the `DecodedAjaxError` too (though I might leave that one for later, as it will require a few changes about the place).

I'm in the habit of using `Either error success` (also a common pattern in Elm), and also I was taking my cue here from existing code in the `pab-client`. But the `Servant.PureScript.Ajax` library uses a `MonadError`, which is ultimately where all this is coming from. So it might make our lives easier if we go along with that, and make our `AppM` a `MonadError` as well. I'll just add a `TODO` comment for now I think.